### ISSUE TITLE
Type inheritance in 0.4.3

### ISSUE THREAD

**Issue by rhelms** [encode/apistar#435]
Is Type inheritance supported in 0.4.3?

For example, I have these two classes:

```
class NbnOrderFeasibilityRequest(types.Type):
    sqId = validators.String()
    nbnLocationId = validators.String()
    tc4DownloadSpeed = validators.Integer()
    tc4UploadSpeed = validators.Integer()
    cpiId = validators.String(default=None)
    ntdId = validators.String(default=None)
    potsInterconnectId = validators.String(default=None)
    transferType = NbnOrderTransferType(default=None)
    localNumberPorting = validators.String(enum=['Yes', 'No'], default=None)
    customerAuthorityDate = validators.Date(default=None),
    appointmentId = validators.String(default=None)
    installationWorkforce = NbnOrderInstallationWorkforceType(default=None)  # TODO conditions
    installationCentralSplitter = validators.String(enum=['Yes', 'No'], default=None)
    exchangePairConnection = validators.String(enum=['No'], default=None)
    ntdPortId = validators.Integer(default=None)
    batteryBackup = validators.String(enum=['Yes', 'No'], default=None)

class NbnOrderRequest(NbnOrderFeasibilityRequest):
    accessSeekerRef = AccessSeekerRef(default=None)
```

When I instantiate an NbnOrderRequest, only accessSeekerRef seems to be available.

    data = {
        'sqId': 'SQ00001',
        'nbnLocationId': 'LOC00001',
        'tc4DownloadSpeed': 10,
        'tc4UploadSpeed': 10
    }
    request = NbnOrderRequest(data)
    assert {'accessSeekerRef': None) == dict(data) # Works, and it shouldn't

I'm guessing there's something that needs to happen in the __new__ operation of TypeMetaclass to get fields/attributes from the parent class, but I just wanted to check if this is supposed to be possible.

**Comment by tomchristie** [comment]
It’s not supported yet, would need a little work on the metaclass. Yes I’d like to see it added. If anyone wants to take it on it’d be worth looking at the metaclass implementation of Django REST framework ModelSerializer

**Comment by tomchristie** [comment]
Also worth noting. In JSON schema we’ll want to map these to allOf cases, so we can included the child and parent definitions. (Tho only works in the case of extending fields, not overriding them)

**Comment by rhelms** [comment]
I'm having a crack at this now.
I've duplicated most tests in test_types.py to make sure the expected operations work with an extended Type.
The bit of code added was very similar to the SerializerMetaclass, except that we're plucking the properties out of the validator on the base class as follows:

        for base in reversed(bases):
            if hasattr(base, 'validator'):
                properties = [
                    (key, base.validator.properties[key]) for key
                    in base.validator.properties
                    if key not in attrs
                ] + properties

The last, and probably, trickiest bit is how this should be presented in the JSON Schema. It's tricky for me because I'm not all that familiar with the JSON Schema, though I'd like to write the test first, and then modify the code to meet the requirement.

Given the existing `Product` type used in the testing, I've extended that type to create a `ReviewedProduct`, which added a `reviewer` property as follows:

```
class ReviewedProduct(Product):
    reviewer = validators.String(max_length=20)
```

My current test for an extended type passes, though the generated structure is flat:

```
def test_extended_as_jsonschema():
    struct = encode_jsonschema(ReviewedProduct, to_data_structure=True)
    assert struct == {
        "$ref": "#/definitions/ReviewedProduct",
        "definitions": {
            "ReviewedProduct": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "maxLength": 10
                    },
                    "rating": {
                        "type": "integer",
                        "minimum": 0,
                        "maximum": 100,
                        "default": None,
                        "nullable": True
                    },
                    "created": {
                        "type": "string",
                        "format": "datetime"
                    },
                    "reviewer": {
                        "type": "string",
                        "maxLength": 20
                    }
                },
                "required": [
                    "name",
                    "created",
                    "reviewer"
                ]
            }
        }
    }
```

At a guess, it should be popping out something like the following:

```
def test_extended_as_jsonschema():
    struct = encode_jsonschema(ReviewedProduct, to_data_structure=True)
    assert struct == {
        "$ref": "#/definitions/ReviewedProduct",
        "definitions": {
            "Product": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "maxLength": 10
                    },
                    "rating": {
                        "type": "integer",
                        "minimum": 0,
                        "maximum": 100,
                        "default": None,
                        "nullable": True
                    },
                    "created": {
                        "type": "string",
                        "format": "datetime"
                    }
                },
                "required": [
                    "name",
                    "created"
                ]
            },
            "ReviewedProduct": {
                "allOf": [
                    {
                        "$ref": "#/definitions/Product"
                    },
                    {
                        "properties": {
                            "reviewer": {
                                "type": "string",
                                "maxLength": 20
                            }
                        },
                        "required": [
                            "reviewer"
                        ]
                    }
                ]
            }
        }
    }
```

I've been looking at the examples from https://spacetelescope.github.io/understanding-json-schema/structuring.html and https://spacetelescope.github.io/understanding-json-schema/reference/combining.html#combining.

I'm also keeping a suspicious eye on Object(Validator).additional_properties. Apparently, if the JSON Schema equivalent is False, then you can't extend that type, and by default, the Object validator used on a Type will set this to None. Though I haven't see the class property pass through to the JSON Schema yet.

If anyone has a bit more experience with JSON Schema than me and could tell me if the presented test is on track or not, that would be great.

**Comment by rhelms** [comment]
I've just been having a look at the `allOf` tests in allOf.json, particularly the "allOf with base schema" entry and the following might also fit the bill.

```
def test_extended_as_jsonschema():
    struct = encode_jsonschema(ReviewedProduct, to_data_structure=True)
    assert struct == {
        "$ref": "#/definitions/ReviewedProduct",
        "definitions": {
            "Product": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "maxLength": 10
                    },
                    "rating": {
                        "type": "integer",
                        "minimum": 0,
                        "maximum": 100,
                        "default": None,
                        "nullable": True
                    },
                    "created": {
                        "type": "string",
                        "format": "datetime"
                    }
                },
                "required": [
                    "name",
                    "created"
                ]
            },
            "ReviewedProduct": {
                "properties": {
                    "reviewer": {
                        "type": "string",
                        "maxLength": 20
                    }
                },
                "required": [
                    "reviewer"
                ],
                "allOf": [
                    {
                        "$ref": "#/definitions/Product"
                    }
                ]
            }
        }
    }
```

**Comment by rhelms** [comment]
Looking at the code a bit more, it would appear that there's no catering for any of the combining types (allOf, anyOf, oneOf).

Introducing allOf is probably a bit beyond my expertise at this point, although I'm hoping it would be something along the lines of adding a defintion to JSON_SCHEMA to cater for the `allOf` structure, and then modifying `JSONSchemaCodec.encode_to_data_structure` to recognise that the item is not a direct subclass of types.Type, and to add a definition of the base type to defs[], determining the new properties introduced by item, and add an `allOf` entry to `value` that references the parent. (* and breathe *)

At this point, I'm going to do a pull request that will observe type inheritance in the classes, but JSON Schema will present the extended class as a flat class, but no knowledge of the parent type.

When allOf is introduced into apistar, then the combination relationship can also be implemented for extended Types.

**Comment by tomchristie** [comment]
> At this point, I'm going to do a pull request that will observe type inheritance in the classes, but JSON Schema will present the extended class as a flat class, but no knowledge of the parent type.

Yup. Approaching this incrementally makes sense.

**Comment by tomchristie** [comment]
Closed via #444. Thanks!
